### PR TITLE
Feature/provide gps information from ausdom a261 dash cam to orb slam2

### PR DESCRIPTION
**Pull request by apattnaik0721013**
@paul-michalik 
- #78 is addressed.
- collect gps information from video.
- map gps information with frame.
- provide encapsulated data (image, timestamp, gps data) to slam.

### CODE REVIEW

**Comment by apattnaik0721013** [comment]
@paul-michalik 
- format of gps data present in video subtitle is : Dialogue:0:00:46.00,0:00:47.00,Default,,0,0,0,,A,071017,160331.304,4815.1500,N,01138.2512,E,32.80,+00.00,+00.00,+00.00;\N
- so it is parsed and stored in vector (':' separate).
gps latitudinal and longitudinal value is collected from index 12 and 14 of the gps data
- class gps_reader is added in ds_dashcam.cc which will extract gps_information from video file.
- for synchronization of frame, frame time is matched with gps frame's pts time / gps frame's frame_duration time.

**Comment by apattnaik0721013** [comment]
@paul-michalik 
Please review the changes.

File: ext/ds_dashcam.cc [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+#include <boost/algorithm/string.hpp>
+
+using namespace std;
+
+namespace ORB_SLAM2 {
+	namespace ext {
+		enum class gps_offset { long_val = 12, long_direction, lat_val, lat_direction };
+
+		class gps_reader
+		{
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: Please put the opening parenthesis on the same line as the class name. Also set up your editor to use 4 whitespaces for indentation.

File: ext/ds_dashcam.cc [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+#include <libavformat/avformat.h>
+#include <libswscale/swscale.h>
+}
+#include <boost/algorithm/string.hpp>
+
+using namespace std;
+
+namespace ORB_SLAM2 {
+	namespace ext {
+		enum class gps_offset { long_val = 12, long_direction, lat_val, lat_direction };
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: Add newlines after enumeration values.

File: ext/ds_dashcam.cc [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+		class gps_reader
+		{
+			std::string				_video_file;
+
+		public:
+			gps_reader(std::string video_file_):_video_file(video_file_)
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: "std::move(video_file_)"

File: ext/ds_dashcam.cc [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+			{
+			}
+
+			//convert gps value(degree minutes and second(dms) to degree decimal(dd))
+			double convert_gps_to_dec(std::string gps_str)
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: Here use "std::string const&" since the value is read only AND isn't copied

File: ext/ds_dashcam.cc [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+			{
+				double gps_val = stod(gps_str);
+				int degree_val = (int)gps_val / 100;
+				double minutes_val = gps_val - (degree_val * 100);
+				double deg_dec_val = degree_val + (minutes_val / 60);
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: Don't create a named value, return the result of the expression immediately.

File: ext/ds_dashcam.cc [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+		public:
+			gps_reader(std::string video_file_):_video_file(video_file_)
+			{
+			}
+
+			//convert gps value(degree minutes and second(dms) to degree decimal(dd))
+			double convert_gps_to_dec(std::string gps_str)
+			{
+				double gps_val = stod(gps_str);
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: Use 'std::stod" from "<string>". You haven't "using namespace std", have you? 😲

File: ext/ds_dashcam.cc [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+				int degree_val = (int)gps_val / 100;
```
Comment (paul-michalik) [inline ext/ds_dashcam.cc]: This smells. Can you cast to `int`. What range of values are we talking about?

**Comment by apattnaik0721013** [comment]
@paul-michalik 
i have addressed all the review comments.

**Comment by paul-michalik** [comment]
Please change the function parameters which are read only and not copied in the function body to "const &". The parameters which are copied should be passed by value and "std::move"-ed to the target!

**Comment by paul-michalik** [comment]
Of course that only applies to non trivial types such as "std::string".

**Comment by apattnaik0721013** [comment]
@paul-michalik 
i have addressed your review comments.

**Comment by apattnaik0721013** [comment]
@paul-michalik 
- please review this branch 
- we can extend data source functionalities in its parent branch

**Comment by paul-michalik** [comment]
OK, so you merged this. in the meantime, there was a pending review which I haven't submitted. Anyways, let's take this as granted and work on #68.

**Pull request merged**